samples/extensions/open_cl_common/open_cl_utils.cpp

```diff
@@ -0,0 +1,133 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @todo: move to sample unit
+
+#include "open_cl_utils.h"
+
+#if defined(__linux__)
+#	include <dlfcn.h>
+#elif defined(_WIN32)
+#	include <windows.h>
+#endif
+#include <string>
+#include <vector>
+
+#define OPENCL_EXPORTED_FUNCTION(func_name) std::function<decltype(func_name)> func_name##_ptr = nullptr
+#define OPENCL_EXPORTED_EXTENSION_FUNCTION(func_name) std::function<decltype(func_name)> func_name##_ptr = nullptr
+#include "open_cl_functions.inl"
+
+static void *handle = nullptr;
+
+cl_platform_id load_opencl()
+{
+	// Check if OpenCL is supported by trying to load the OpenCL library and getting a valid function pointer
+	bool openCLAvailable = false;
+
+#if defined(__ANDROID__)
+	// Try to find the OepenCL library on one of the following paths
+	static const char *libraryPaths[] = {
+	    // Generic
+	    "/system/vendor/lib64/libOpenCL.so",
+	    "/system/lib64/libOpenCL.so",
+	    "/system/vendor/lib/libOpenCL.so",
+	    "/system/lib/libOpenCL.so",
+	    // ARM Mali
+	    "/system/vendor/lib64/egl/libGLES_mali.so",
+	    "/system/lib64/egl/libGLES_mali.so",
+	    "/system/vendor/lib/egl/libGLES_mali.so",
+	    "/system/lib/egl/libGLES_mali.so",
+	    // PowerVR
+	    "/system/vendor/lib64/libPVROCL.so",
+	    "/system/lib64/libPVROCL.so",
+	    "/system/vendor/lib/libPVROCL.so",
+	    "/system/lib/libPVROCL.so"};
+	for (auto libraryPath : libraryPaths)
+	{
+		handle = dlopen(libraryPath, RTLD_LAZY);
+		if (handle)
+		{
+			break;
+		}
+	}
+#elif defined(__linux__)
+	// Try to find the OepenCL library on one of the following paths
+	static const char *libraryPaths[] = {
+	    "libOpenCL.so",
+	    "/usr/lib/libOpenCL.so",
+	    "/usr/local/lib/libOpenCL.so",
+	    "/usr/local/lib/libpocl.so",
+	    "/usr/lib64/libOpenCL.so",
+	    "/usr/lib32/libOpenCL.so",
+	    "libOpenCL.so.1",
+	    "/usr/lib/libOpenCL.so.1",
+	    "/usr/local/lib/libOpenCL.so.1",
+	    "/usr/local/lib/libpocl.so.1",
+	    "/usr/lib64/libOpenCL.so.1",
+	    "/usr/lib32/libOpenCL.so.1"};
+	for (auto libraryPath : libraryPaths)
+	{
+		handle = dlopen(libraryPath, RTLD_LAZY);
+		if (handle)
+		{
+			break;
+		}
+	}
+#elif defined(_WIN32)
+	handle = LoadLibraryA("OpenCL.dll");
+	if (handle)
+	{
+		char libPath[MAX_PATH] = {0};
+		GetModuleFileNameA(static_cast<HMODULE>(handle), libPath, sizeof(libPath));
+	}
+#endif
+
+	if (handle == nullptr)
+	{
+		return nullptr;
+	}
+
+#if defined(_WIN32)
+#	define OPENCL_EXPORTED_FUNCTION(func_name) func_name##_ptr = reinterpret_cast<decltype(func_name) *>(GetProcAddress((HMODULE) handle, #func_name))
+#else
+#	define OPENCL_EXPORTED_FUNCTION(func_name) func_name##_ptr = reinterpret_cast<decltype(func_name) *>(dlsym(handle, #func_name))
+#endif
+#include "open_cl_functions.inl"
+
+	cl_platform_id platform_id = nullptr;
+	cl_uint        num_platforms;
+	clGetPlatformIDs_ptr(1, &platform_id, &num_platforms);
```

Comment:
This works by accident only. Vulkan and OpenCL might (and will) enumerate the devices differently. That is, to get the OpenCL platform corresponding to the selected Vulkan physical device, you should select the OpenCL platform with the same UUID.

Reply:
That's a valuable information. Will change it!

Reply:
And same as above: That's code from ARM, so I probably also need to rework their sample :/

Reply:
Though the question I'm asking myself: Does it really matter? This (and the other sample) are about Vulkan and OpenCL interop. And should it matter if you do it on the same device or two different devices?

Reply:
To get the UUID of the OpenCL device, you need the [cl_khr_device_uuid](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_device_uuid) extension. Unfortunately, the extension support is checked at a later point in time in your sample.

Reply:
Yeah, I know. I'm already reworking that part ;)

Reply:
I have reworked how this works and now try to find a matching OpenCL device using the UUID from Vulkan. I also updated the documentation. Feedback is appreciated :)